Remove version dependence from codegen

fbs opened the pull request:
sorry for the noisy PR. All the magic happens in `codegen/llvm/common.h`

----

The codegen tests are currently quite tedious to update as the "source"
and "result" are defined in the same file which makes it hard to
automatically update.

This commit changes that by:

- Moving the "result" to its own file, `TEST_NAME.ll`
- Introducing a rewriter that tries to automatically fix the LLVM
  version differences
- As rewriting doesn't always work it supports loading results from
  `TEST_NAME_LLVM-X.ll` these will have to be maintained manually but
  at the moment there are only a few (only LLVM 5)

fbs commented:
This  hopefully makes  adding llvm-9 codegen tests a  lot easier too. I will give that a  try  once (if) this gets merged

fbs commented on tests/codegen/common.h:
```diff
@@ -18,17 +18,151 @@ namespace bpftrace {
+  {
+    if (*it == "align")
+    {
+      if (alignment == "")
```
This is a bit of a hack as we have a few tests with a funky address space, so only the value after the first align is valid:

```
call void @llvm.memcpy.p0i8.p64i8.i64(i8* nonnull align 1 %1, i8 addrspace(64)* align 536870912 null, i64 32, i1 false)
```

danobi replied:
Maybe add a comment mentioning it

danobi commented on tests/codegen/common.h:
```diff
@@ -18,17 +18,151 @@ namespace bpftrace {
+  auto parts = split_string(line, ' ');
+  std::stringstream buf;
+  std::string alignment = "";
+  for (std::vector<std::string>::iterator it = parts.begin(); it != parts.end();
```
```suggestion
  for (auto it = parts.begin(); it != parts.end();
```

danobi commented on tests/codegen/common.h:
```diff
@@ -18,17 +18,151 @@ namespace bpftrace {
+  // clang-format on
+  auto parts = split_string(line, ' ');
+  std::stringstream buf;
+  std::string alignment = "";
```
Default initializes to empty
```suggestion
  std::string alignment;
```

danobi commented on tests/codegen/common.h:
```diff
@@ -18,17 +18,151 @@ namespace bpftrace {
+  {
+    if (*it == "align")
+    {
+      if (alignment == "")
```
Nit
```suggestion
      if (alignment.empty())
```

danobi commented on tests/codegen/common.h:
```diff
@@ -18,17 +18,151 @@ namespace bpftrace {
+  }
+  return buf.str();
+#else
+  return std::string(line);
```
```suggestion
  return line;
```

danobi commented on tests/codegen/common.h:
```diff
@@ -18,17 +18,151 @@ namespace bpftrace {
+  return std::string("declare void @llvm.memcpy.p0i8.p0i8.i64(i8* nocapture "
+                     "writeonly, i8* nocapture readonly, i64, i32, i1) #1");
+#else
+  return std::string(line);
```
```suggestion
  return line;
```

danobi commented on tests/codegen/common.h:
```diff
@@ -18,17 +18,151 @@ namespace bpftrace {
+static std::string rewrite(const std::string &ir)
+{
+  std::stringstream buf;
+  for (auto &line : split_string(ir, '\n'))
```
```suggestion
  for (const auto &line : split_string(ir, '\n'))
```

fbs commented:
@danobi  thanks for the review. I've updated it to have the test runner update the codegen when `BPFTRACE_UPDATE_TESTS` is set to `1` in the env.  Still not the best user experience when updating tests but it should be a lot better than is it now

danobi commented:
Oh, also please add documentation. Maybe in the README.md in tests/

fbs commented:
@danobi / @ajor  I've added docs and added  LLVM9 support (which turned out quite trivial). Can you take a look at last 2 commits? 

https://github.com/iovisor/bpftrace/pull/1108/commits/6036494814b6a6bbd3fc3eb4474f0f763f87c33f

fbs commented:
@mmisono  any objection to me merging this or would you rather wait until #1104 is merged as it impacts  your work? It's blocking #1105 and testing on fedora >=31 systems.

mmisono commented:
@fbs
I'm glad if #1104 is merged first, but not sure how long it takes, so please go ahead. This and #1105 are important changes and I also want them to be merged.

By the way, I think it is useful if bpftrace has an option to save LLVM IR to a file like [this --elf option](https://github.com/iovisor/bpftrace/pull/871/commits/c97ce70aada26e63b1a49fe2c36c2475b83).

fbs commented:
> @fbs
> I'm glad if #1104 is merged first, but not sure how long it takes, so please go ahead. This and #1105 are important changes and I also want them to be merged.

I'll hold it off for the weekend then :)

> By the way, I think it is useful if bpftrace has an option to save LLVM IR to a file like [this --elf option](https://github.com/iovisor/bpftrace/pull/871/commits/c97ce70aada26e63b1a49fe2c36c2475b83).

I had something like this in #1066 but there is some discussion still about how to name the flags.  I got a WIP somewhere that adds `--emit-<ir,opt-ir,elf> [filename]` somewhere, i'll try and clean that up and open it as a rfc for discussion :)
